### PR TITLE
HTTP2: Implement buffering of H2 stream data and trailers.

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

~~DO NOT SUBMIT UNTIL FLIPPING THIS TO BE OFF BY DEFAULT! -> It is now off by default but parameterized in various tests to avoid bitrot.~~ This is off by default, but is parameterized in many tests to avoid bitrot. For clarity, changes to filters that made assumptions on the data flow that is broken by this will have a follow up PR parametrizing them.

Commit Message:  Implement buffering of H2 stream data and trailers.
Additional Description: See related issue, this is a first iteration of this and has a few issues such as potential for stream starvation since we process all the buffered data, hence one stream on a connection could end up hogging the connection.
Risk Level: High (but this is off by default)
Testing: Added Unit / Integration tests
Docs Changes: Included
Release Notes: Included
Platform Specific Features: N/A
Related Issue #19127

### CODE REVIEW

**Comment by KBaichoo** [comment]
/assign @alyssawilk

Can you take a look at this? Thanks

**Comment by alyssawilk** [comment]
Sure, but I'm inclined to get a first pass from @antoniovicente

**Comment by alyssawilk** [comment]
oh I think this is getting lost due to draft status.  mind marking as ready for review even ifit's not ready for submit?

File: source/common/http/http2/codec_impl.cc [hunk 1]
```diff
@@ -349,6 +354,36 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
   }
 }
 
+void ConnectionImpl::StreamImpl::processBufferedData(bool stream_being_destroyed) {
+  ENVOY_CONN_LOG(debug, "Stream {} processing buffered data.", parent_.connection_, stream_id_);
+
+  auto next_stage = stream_manager_.getNextStage();
+  stream_manager_.flush_all_data_ = stream_being_destroyed;
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:361]: maybe some comments?  on first pass my I wonder why we'd flush all data if the stream is going away?  seems like a lot of work for a dying stream

File: source/common/http/http2/codec_impl.cc [hunk 2]
```diff
@@ -349,6 +354,36 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
...
+  bool continue_processing = stream_being_destroyed || !buffersOverrun();
+  while (continue_processing && next_stage.has_value()) {
+    switch (next_stage.value()) {
+    case BackedUpStreamManager::Stage::Body:
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:365]: this feels like it was written for the headers/body/trailers case but I think we can maybe simplify if it's just body and data.  I'd be inclined to 
while (data && coninue) {
  do data
}
if (continue) {
do trailers
}
WDYT?

File: source/common/http/http2/codec_impl.cc [hunk 3]
```diff
@@ -349,6 +354,36 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
...
+      if (stream_manager_.data_end_stream_) {
+        remote_end_stream_ = true; // restore the deferred end_stream from the remote.
+      } else if (stream_manager_.trailers_buffered_) {
+        // Avoid inversion in the case where we saw trailers, acquiring the
+        // remote_end_stream_ being set to true, but the trailers ended up being
+        // buffered. The buffered decodeTrailers call will restore the
+        // remote_end_stream_.
+        remote_end_stream_ = false;
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:373]: I don't love overloading remote_end_stream.  Can we put this logic in decodeData as send_end_stream = got_end_stream + no_trailers?

File: source/common/http/http2/codec_impl.cc [hunk 4]
```diff
@@ -374,17 +419,57 @@ void ConnectionImpl::StreamImpl::readDisable(bool disable) {
...
+  // Any buffered data will be consumed.
+  stream_manager_.body_buffered_ = false;
+
+  // It's possible that we are waiting to send a deferred reset, so only raise data if local
+  // is not complete.
+  if (!deferred_reset_) {
+    decoder().decodeData(*pending_recv_data_, remote_end_stream_);
+  }
+
+  pending_recv_data_->drain(pending_recv_data_->length());
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:472]: I thought we were going to dump in chunks, whatever the default read limit is?

**Review by alyssawilk — COMMENTED** [review]
Thanks for tackling this one - it's going to be such a win!

**Review by KBaichoo — COMMENTED** [review]
Thanks for a first pass, still working on two failing tests
```
//test/extensions/filters/http/grpc_json_transcoder:grpc_json_transcoder_integration_test
//test/extensions/filters/http/ext_proc:streaming_integration_test
```

That seem to be related to the workaround of pushing data through if we see onStreamClose.

File: source/common/http/http2/codec_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/http/http2/codec_impl.cc:364]: I guess I'm confused how anything is closing the stream if the codec doesn't communicate to other layers that the stream is complete?

File: source/common/http/http2/codec_impl.cc (hunk 4 above)
Comment (alyssawilk) [reply source/common/http/http2/codec_impl.cc:459]: Ok, let's definitely add a TODO here then.  I think delaying this for a bit is fine but I'd definitely like to sort out the stream flush-close issue before we get too deep into the weeds

Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc:471]: added todo

File: source/common/http/http2/codec_impl.cc (hunk 1 above)
Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc]: As discussed offline, ended up achieving this by buffering onStreamClose call that way we don't have this hack where we flush all the data through.

**Comment by alyssawilk** [comment]
@KBaichoo this ready for another pass?

**Comment by KBaichoo** [comment]
yes

File: source/common/http/http2/codec_impl.cc [hunk 5]
```diff
@@ -166,6 +166,11 @@ ConnectionImpl::StreamImpl::StreamImpl(ConnectionImpl& parent, uint32_t buffer_l
 }
 
 void ConnectionImpl::StreamImpl::destroy() {
+  // Cancel any pending buffered data callback for the stream.
+  if (process_buffered_data_callback_ != nullptr && process_buffered_data_callback_->enabled()) {
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:170]: would it make sense to cancel in the data callback destructor and just reset() here?

File: source/common/http/http2/codec_impl.cc [hunk 6]
```diff
@@ -349,6 +354,39 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
   }
 }
 
+void ConnectionImpl::StreamImpl::processBufferedData() {
+  ENVOY_CONN_LOG(debug, "Stream {} processing buffered data.", parent_.connection_, stream_id_);
+
+  // We should stop processing buffered data if either
+  // 1) Buffers become overrun
+  // 2) The stream ends up getting reset
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:362]: What if sending data results in an error and there's a local reply generated instead? 
Or does downstream local reply reset unfinished upstream stream so we're Ok just using reset?  I think we'll want an integration test for both downstream reset and downstream local reply either way.

File: source/common/http/http2/codec_impl.cc [hunk 7]
```diff
@@ -349,6 +354,39 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
...
+  // Both of these can end up changing as a result of processing buffered data.
+  bool continue_processing = !buffersOverrun() && !reset_reason_.has_value();
+
+  if (stream_manager_.body_buffered_ && continue_processing) {
+    if (stream_manager_.data_end_stream_) {
+      remote_end_stream_ = true; // restore the deferred end_stream from the remote.
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:368]: I'd be inclined to keep remote_end_stream as it was, but change the call sites which use it to use a utility function sendEndStream() which does the needful.  Alternately you could name remote_end_stream to send_remote_end_stream but having received a remote end stream and having this false is likely to cause debug confusion.

File: source/common/http/http2/codec_impl.cc [hunk 8]
```diff
@@ -349,6 +354,39 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
...
+    }
+    decodeData();
+  }
+
+  continue_processing = !buffersOverrun() && !reset_reason_.has_value();
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:373]: worth making this a helper function, especially if we track local reply?

File: source/common/http/http2/codec_impl.cc [hunk 9]
```diff
@@ -374,17 +422,53 @@ void ConnectionImpl::StreamImpl::readDisable(bool disable) {
...
-  ENVOY_CONN_LOG(debug, "recv buffer over limit ", parent_.connection_);
-  ASSERT(!pending_receive_buffer_high_watermark_called_);
-  pending_receive_buffer_high_watermark_called_ = true;
-  readDisable(true);
+  if (!Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.defer_processing_backedup_streams")) {
+    ENVOY_CONN_LOG(debug, "recv buffer over limit ", parent_.connection_);
+    ASSERT(!pending_receive_buffer_high_watermark_called_);
+    pending_receive_buffer_high_watermark_called_ = true;
+    readDisable(true);
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:430]: wait, are we not read disabling in the new path? why?

File: source/common/http/http2/codec_impl.cc [hunk 10]
```diff
@@ -408,7 +492,32 @@ void ConnectionImpl::ClientStreamImpl::decodeHeaders() {
...
 }
 
+bool ConnectionImpl::StreamImpl::maybeDeferDecodeTrailers() {
+  // Buffer trailers if we're deferring processing and not flushing all data
+  // through and either
+  // 1) Buffers are overrun
+  // 2) There's buffered body which should get processed before these trailers
+  //    to avoid losing data.
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.defer_processing_backedup_streams") &&
```
Comment (alyssawilk) [inline source/common/http/http2/codec_impl.cc:502]: So if this flag was true, we buffered some body, we flip the flag false, and get trailers, what happens?
Please run through this PR and make sure it's safe to flip the flag from true -> false or false -> true at any point, or else latch it on stream creation for simplicity :-)

**Review by alyssawilk — COMMENTED** [review]
I think clang tidy needs a look?  else if it's flaky, bonus point for filing an issue :-)

File: source/common/http/http2/codec_impl.cc (hunk 6 above)
Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc]: I think it's the latter, for example in the grpc json transcoder test, if we're above a threshold in the filter we end up sending a local reply as a result of processing this buffered data. Will add some other tests to explicitly test this though.

File: source/common/http/http2/codec_impl.cc (hunk 7 above)
Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc]: great idea, done :)

**Comment by alyssawilk** [comment]
ping @yanavlasov @antoniovicente

File: source/common/http/http2/codec_impl.cc [hunk 11]
```diff
@@ -349,6 +354,28 @@ void ConnectionImpl::StreamImpl::encodeMetadata(const MetadataMapVector& metadat
...
+  }
+
+  // Activate the buffered onStreamClose if we've drained all the buffered data,
+  // or we've gotten a reset.
+  if (stream_manager_.buffered_on_stream_close_ &&
+      (!stream_manager_.hasBufferedData() || reset_reason_.has_value())) {
+    ENVOY_CONN_LOG(debug, "invoking  onStreamClose for stream: {} via processBufferedData",
+                   parent_.connection_, stream_id_);
+    // We only buffer the onStreamClose if we had no errors.
+    parent_.onStreamClose(this, 0);
```
Comment (antoniovicente) [inline source/common/http/http2/codec_impl.cc:375]: There is a call to parent_.onStreamClose when reset_reason_ is set in ConnectionImpl::StreamImpl::resetStream.  Can the check for reset_reason_.has_value() result in this getting called twice?  I think the answer is no, because buffered_on_stream_close_ is cleared by parent_.onStreamClose, but also implies that the reset_reason_.has_value() check is not needed here.

File: source/common/http/http2/codec_impl.cc [hunk 12]
```diff
@@ -1265,6 +1380,14 @@ int ConnectionImpl::onStreamClose(int32_t stream_id, uint32_t error_code) {
       }
 
       stream->runResetCallbacks(reason);
+    } else if (!stream->reset_reason_.has_value() && stream->stream_manager_.hasBufferedData()) {
```
Comment (antoniovicente) [inline source/common/http/http2/codec_impl.cc:1383]: Could you explain the sequence of events that can lead you to get here?  It's unclear to me why this appears relevant in cases where defer_processing_backedup_streams_ is false.

File: source/common/http/http2/codec_impl.h [hunk 13]
```diff
@@ -341,13 +352,43 @@ class ConnectionImpl : public virtual Connection,
...
+     */
+    struct BufferedStreamManager {
+      bool body_buffered_{false};
+      bool trailers_buffered_{false};
+
+      // We received a call to onStreamClose for the stream, but deferred it
+      // as the stream had pending data to process and the stream was not reset.
+      bool buffered_on_stream_close_{false};
+
+      bool hasBufferedData() const { return body_buffered_ || trailers_buffered_; }
```
Comment (antoniovicente) [inline source/common/http/http2/codec_impl.h:369]: nit: hasBufferedBodyOrTrailers

File: source/common/http/http2/codec_impl.h [hunk 14]
```diff
@@ -341,13 +352,43 @@ class ConnectionImpl : public virtual Connection,
...
+      // 1) Buffers become overrun
+      // 2) The stream ends up getting reset
+      // Both of these can end up changing as a result of processing buffered data.
+      return !buffersOverrun() && !reset_reason_.has_value();
+    }
+
+    // Avoid inversion in the case where we saw trailers, acquiring the
+    // remote_end_stream_ being set to true, but the trailers ended up being
+    // buffered.
+    bool sendEndStream() const { return remote_end_stream_ && !stream_manager_.trailers_buffered_; }
```
Comment (antoniovicente) [inline source/common/http/http2/codec_impl.h:391]: What about !stream_manager_.body_buffered_ ?

This seems relevant in ConnectionImpl::ClientStreamImpl::decodeHeaders(), the other 2 calls I can understand at least while you're moving the full body through the filter pipeline in 1 go.

File: source/common/runtime/runtime_features.cc [hunk 15]
```diff
@@ -61,6 +61,7 @@ constexpr const char* runtime_features[] = {
     "envoy.reloadable_features.conn_pool_delete_when_idle",
     "envoy.reloadable_features.correct_scheme_and_xfp",
     "envoy.reloadable_features.correctly_validate_alpn",
+    "envoy.reloadable_features.defer_processing_backedup_streams",
```
Comment (antoniovicente) [inline source/common/runtime/runtime_features.cc:64]: Aren't these enabled by default?  The release note says that this behavior is disabled by default.

File: docs/root/configuration/http/http_conn_man/stats.rst [hunk 16]
```diff
@@ -173,6 +173,7 @@ On the upstream side all http2 statistics are rooted at *cluster.<name>.http2.*
    keepalive_timeout, Counter, Total number of connections closed due to :ref:`keepalive timeout <envoy_v3_api_field_config.core.v3.KeepaliveSettings.timeout>`
    streams_active, Gauge, Active streams as observed by the codec
    pending_send_bytes, Gauge, Currently buffered body data in bytes waiting to be written when stream/connection window is opened.
+   deferred_stream_close, Gauge, Number of streams such that we deferred closing the stream due to it buffering data to be consumed by its corresponding stream.
```
Comment (antoniovicente) [inline docs/root/configuration/http/http_conn_man/stats.rst:176]: I'm having trouble parsing this doc string.

File: test/integration/multiplexed_upstream_integration_test.cc [hunk 17]
```diff
@@ -342,6 +342,12 @@ TEST_P(MultiplexedUpstreamIntegrationTest, ManyLargeSimultaneousRequestWithRando
   if (upstreamProtocol() == Http::CodecType::HTTP3) {
     return;
   }
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.defer_processing_backedup_streams")) {
+    // We end up in a state where the filter pauses the stream, but never
+    // unpauses it as we buffer data at the receiving codec.
+    return;
```
Comment (antoniovicente) [inline test/integration/multiplexed_upstream_integration_test.cc:349]: Can we have at least a TODO to get this test fixed?  What would be required to make it work?

File: test/integration/http2_flood_integration_test.cc [hunk 18]
```diff
@@ -595,6 +595,15 @@ TEST_P(Http2FloodMitigationTest, Trailers) {
 // Verify flood detection by the WINDOW_UPDATE frame when a decoder filter is resuming reading from
 // the downstream via DecoderFilterBelowWriteBufferLowWatermark.
 TEST_P(Http2FloodMitigationTest, WindowUpdateOnLowWatermarkFlood) {
+  // This test depends on data flowing through a backed up stream eagerly (e.g. the
+  // backpressure-filter triggers above watermark when it receives headers from
+  // the downstream, and only goes below watermark if the response body has
+  // passed through the filter.). With defer processing of backed up streams however
+  // the data won't be eagerly processed as the stream is backed up.
```
Comment (antoniovicente) [inline test/integration/http2_flood_integration_test.cc:602]: at least add a TODO to remove this test when removing this feature tag.

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 19]
```diff
@@ -355,6 +313,73 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
...
+
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.defer_processing_backedup_streams") &&
+        state.queueOverHighLimit()) {
+      // We just transitioned to queue over high limit, which will then cause
+      // the receiving codec to buffer the data rather than send it to us.
+      // Prior, we'd rely on eager processing to kick off sending the data to
+      // the external processor, instead we kick it off here.
+      Buffer::OwnedImpl empty_buffer{};
+      return processBufferedPartial(state, empty_buffer, end_stream);
```
Comment (antoniovicente) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:374]: Are some adjustments to end_stream required in this case?

Comment (antoniovicente) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:374]: What prevents this code from going into an infinite loop?  What is the expected return value of this recursive call?

**Review by antoniovicente — COMMENTED** [review]
Very interesting change.  Seems on the right track in many ways, some comments below.

File: source/common/http/http2/codec_impl.cc (hunk 11 above)
Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc:377]: Good reading, sir. yes as you said buffered_on_stream_close_ prevents us from closing the stream twice. The check of `reset_reason_.has_value()` here in not needed.

File: source/common/http/http2/codec_impl.cc (hunk 9 above)
Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc:424]: Prior it seems to never be called: https://github.com/envoyproxy/envoy/pull/19447/files/82147f0ce2cf7deff38beaa1af519dcc7cf9116d#diff-d36bad9503815c85f0a45bdd1b7342a73112349b6167e8170af677e0b9225990R329-R336

since we'd always eagerly pass the data through.  To preserve that behavior with deferred processing, and avoid getting stuck we disable it's watermarks.

**Comment by alyssawilk** [comment]
Looks like tests are still unhappy, but when they're sorted please PTAL @antoniovicente @yanavlasov

**Comment by KBaichoo** [comment]
I think clang tidy error needs a no-lint and that the code coverage should go back up when I flip this to be default off as with it default on the watermark callback for pending_recv_buffer is disabled, see: https://storage.googleapis.com/envoy-pr/41f582f/coverage/source/common/http/http2/codec_impl.cc.gcov.html

Is this at a state I should flip this to be default off? @alyssawilk

File: source/common/http/http2/codec_impl.cc (hunk 9 above)
Comment (antoniovicente) [reply source/common/http/http2/codec_impl.cc:424]: My worry is that after this change MarkDataConsumedForStream will be called early, effectively breaking flow control feedback.  Should you update the documentation of how H2 flow control works to reflect changes in this PR?

File: source/common/http/http2/codec_impl.h (hunk 14 above)
Comment (antoniovicente) [reply source/common/http/http2/codec_impl.h:391]: Worth an ASSERT(!stream_manager_.body_buffered_) ?

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 19 above)
Comment (antoniovicente) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:374]: Should you be passing false instead of end_stream?

File: source/common/http/http2/codec_impl.cc (hunk 9 above)
Comment (yanavlasov) [reply source/common/http/http2/codec_impl.cc:424]: I **think** the only case where this is called is when a filter adds more bytes to the received data in the decodeData()/decodeHeaders() call (i.e. decompressor filter). And I **think** the only outcome of this is reset stream.

I will look into this more, but I think whatever this callback is doing is largely a noop. I left a [comment](https://github.com/envoyproxy/envoy/blob/d3ecdce968830330fbffcfc492d28c212c509acf/source/common/http/http2/codec_impl.h#L324) a while ago, which is actually not quite correct, as it does not mention the edge case above.

File: source/common/runtime/runtime_features.cc (hunk 15 above)
Comment (yanavlasov) [reply source/common/runtime/runtime_features.cc:66]: I think we should add a parameter to some of existing test suites. codec_impl_test.cc as well, some integration tests (while adding more shards to avoid it timing out). In this case you do not need to worry about the default value of this flag.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
First pass complete. Will finish later today.

/wait

**Comment by KBaichoo** [comment]
ping @antoniovicente @yanavlasov  anything else?

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

**Review by yanavlasov — COMMENTED** [review]
LGTM from me. Just need to get CI green

/wait

File: test/integration/buffer_accounting_integration_test.cc [hunk 20]
```diff
@@ -704,4 +716,263 @@ TEST_P(Http2OverloadManagerIntegrationTest, CanResetStreamIfEnvoyLevelStreamEnde
...
+  write_matcher_->setResumeWrites();
+
+  ASSERT_TRUE(response_decoder->waitForReset());
+  ASSERT_TRUE(upstream_request->complete());
+  test_server_->waitForGaugeEq("cluster.cluster_0.http2.deferred_stream_close", 0);
+  test_server_->waitForCounterEq("cluster.cluster_0.upstream_flow_control_resumed_reading_total",
+                                 1);
+}
+
+TEST_P(Http2DeferredProcessingIntegrationTest,
```
Comment (alyssawilk) [inline test/integration/buffer_accounting_integration_test.cc:931]: So I'd like to see 2 tests in addition to these integration test, and the two test are 
while data is backed up in the h2 buffers, some intermediate filter (unzip? but can be totally synthetic) adds enough data to have us go over the watermark limits and send a local response with [error response headers, reset]

I am 100% fine with a TODO for tese tests given the flag is off by default, but filter chain local response has been a common error mode and I want to make sure we handle this before we flip true.

**Review by alyssawilk — COMMENTED** [review]
Yay!  This is so much cleaner!

File: source/common/http/http2/codec_impl.cc (hunk 9 above)
Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc:428]: I think before we'd end up doing the "disable-h2-window-consume, dispatch data, enable-h2-window-consume" in the same callback usually. 

What are your thoughts on making this readDisable as follows: 
1) this is somehow not counted in the the pending data processing case e.g. to avoid getting stuck 
2) this is counted in the `MarkDataConsumed` case in `ConnectionImpl::OnData()` so we don't give the peer window 

This would thus use the overflow / low watermark to inform flow control, but it won't control whether we process received data

File: test/common/http/http2/codec_impl_test.cc [hunk 21]
```diff
@@ -3223,6 +3264,54 @@ TEST_P(Http2CodecImplTest, ConnectTest) {
...
+  EXPECT_CALL(request_decoder_, decodeTrailers_(_)).Times(0);
+  request_encoder_->encodeTrailers(TestRequestTrailerMapImpl{{"trailing", "header"}});
+  driveToCompletion();
+
+  // Now invoke the deferred processing callback.
+  {
+    InSequence seq;
+    EXPECT_CALL(request_decoder_, decodeData(_, false));
+    EXPECT_CALL(request_decoder_, decodeTrailers_(_));
+    process_buffered_data_callback->invokeCallback();
```
Comment (alyssawilk) [inline test/common/http/http2/codec_impl_test.cc:3311]: I think you could do with a bit more unit testing of corner cases here.  basics like back up of body with no fin doesn't send fin on resume.  back up of body with fin and no trailers, etc.

**Review by alyssawilk — COMMENTED** [review]
code fundamentally LGTM modulo that one flow-control-or-envoy-bug change.  Wohoo!

File: test/common/http/http2/codec_impl_test.cc (hunk 21 above)
Comment (KBaichoo) [reply test/common/http/http2/codec_impl_test.cc:3311]: added additional unit test

**Comment by KBaichoo** [comment]
/assign @mattklein123

Mind taking a look as a non-googler to sign off? Thanks!

**Review by alyssawilk — APPROVED** [review]
Cool, once Matt does his pass let's ship and iterate

File: source/docs/flow_control.md [hunk 22]
```diff
@@ -101,6 +101,29 @@ upstream connections, the `readDisable(true)` calls are unwound in
...
+
+* If a HTTP/2 stream is read disabled anywhere we will begin buffering body and
+  trailers in the receiving side of the codec to minimize work done on the
+  stream when it's read disabled. There will be a callback scheduled to drain
+  these when the stream is read enabled, it can also be drained if the stream is
+  read enabled, and the stream is invoked with additional data.
+* The codec's recieve buffer high watermark is still used in consideration for
+  granting peer stream additional window (preserving existing protocol flow
+  control), but no longer read disables the stream, as we could be read disabled
+  elsewhere, buffer data in codec's recieve buffer hitting high watermark and
```
Comment (mattklein123) [inline source/docs/flow_control.md:118]: Per the giant comment thread, my understanding also was that the recv side read disable was never really used, as the data was always sent directly through. I remembering commenting on this eons ago with @alyssawilk originally did the work. I think this is what @yanavlasov was referring to also. Assuming this is the case, can you pedantically clarify this statement just to make it more clear that we were never previously actually relying on this? I think the idea now is that we still do read disable "from the other side" we just now buffer on the origin side vs. the receiving side. Right?

File: test/extensions/filters/http/fault/BUILD [hunk 23]
```diff
@@ -54,6 +54,7 @@ envoy_extension_cc_test(
     name = "fault_filter_integration_test",
     srcs = ["fault_filter_integration_test.cc"],
     extension_names = ["envoy.filters.http.fault"],
+    shard_count = 2,
```
Comment (mattklein123) [inline test/extensions/filters/http/fault/BUILD:57]: nit: are these changes still necessary? Can they be reverted for now for clarity? Same below.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with tiny comments.

/wait

File: source/docs/flow_control.md (hunk 22 above)
Comment (KBaichoo) [reply source/docs/flow_control.md:118]: Clarified as suggested. The idea now is we read disable from the other side (as done prior), but buffer on the receiving side vs the connection output buffer which the data would get serialized to.

**Review by mattklein123 — APPROVED** [review]
Thanks!
